### PR TITLE
feat: implement inference server by using vllm

### PR DESCRIPTION
**Pull request by zhuangqh**
**Reason for Change**:

**Requirements**

- [x] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

**Notes for Reviewers**:

### CODE REVIEW

**Comment by zhuangqh** [comment]
testing code
```
from openai import OpenAI

# Modify OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:5000/v1"

client = OpenAI(
    # defaults to os.environ.get("OPENAI_API_KEY")
    api_key=openai_api_key,
    base_url=openai_api_base,
)

models = client.models.list()
model = models.data[0].id

completion = client.chat.completions.create(
    model=model,
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {
            "role": "user",
            "content": "What is kubernetes?"
        }
    ],
    n=10000,
)

print(completion.choices[0].message)
```

**Comment by zhuangqh** [comment]
link #608

File: presets/inference/text-generation/inference_api_vllm.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        "swap-space": 4,
+        "disable_log_stats": False,
+    }
+    parser.set_defaults(**engine_default_args)
+
+    return parser
+
+
+if __name__ == "__main__":
+    parser = FlexibleArgumentParser(description='vLLM serving server')
```
Comment (zhuangqh) [inline presets/inference/text-generation/inference_api_vllm.py:89]: the same ux as `vllm serve`

File: presets/inference/text-generation/api_spec.json [hunk 2]
```diff
@@ -5,136 +5,144 @@
         "version": "0.1.0"
```
Comment (ishaansehgal99) [inline presets/inference/text-generation/api_spec.json:5]: Create new file for this? Don't overwrite old api's spec

File: presets/inference/text-generation/inference_api.py [hunk 3]
```diff
@@ -153,7 +161,7 @@ def __post_init__(self): # validate parameters
     pipeline_kwargs["torch_dtype"] = args.torch_dtype
 
 pipeline = transformers.pipeline(
-    model_pipeline,
+    task="text-generation",
```
Comment (bangqipropel) [inline presets/inference/text-generation/inference_api.py:164]: If you don't use the original parameter, then consider remove the corresponding parts

Comment (zhuangqh) [reply presets/inference/text-generation/inference_api.py:166]: using in /chat handler

File: presets/tuning/text-generation/requirements.txt [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Core Dependencies
 transformers==4.41.2
-torch==2.2.0
+torch==2.4.0
```
Comment (ishaansehgal99) [inline presets/tuning/text-generation/requirements.txt:3]: Why editing working torch version for fine tuning?

File: presets/inference/text-generation/inference_api_vllm.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    status: str = Field(..., example="Healthy")
+
+@api_server.router.get("/healthz")
+async def health_check(raw_request: Request) -> HealthStatus:
+    """Health check."""
+    try:
+        await asyncio.wait_for(api_server.engine_client(raw_request).check_health(), timeout=1.0)
+        return HealthStatus(status="Healthy")
+    except asyncio.TimeoutError:
+        raise HTTPException(status_code=500, detail="Model not initialized")
```
Comment (ishaansehgal99) [inline presets/inference/text-generation/inference_api_vllm.py:29]: does vllm already come with health endpoint?

**Review by ishaansehgal99 — COMMENTED** [review]
-

File: presets/tuning/text-generation/requirements.txt (hunk 4 above)
Comment (zhuangqh) [reply presets/tuning/text-generation/requirements.txt:3]: By mistake, will fix

File: presets/inference/text-generation/inference_api_vllm.py (hunk 5 above)
Comment (zhuangqh) [reply presets/inference/text-generation/inference_api_vllm.py:29]: Yes, but I don't want to expose this difference to controller part.

File: presets/inference/text-generation/inference_api_vllm.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def make_arg_parser(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
+    local_rank = int(os.environ.get("LOCAL_RANK",
+                                    0))  # Default to 0 if not set
+    port = 5000 + local_rank  # Adjust port based on local rank
+
+    server_default_args = {
+        "disable-frontend-multiprocessing": False,
```
Comment (ishaansehgal99) [inline presets/inference/text-generation/inference_api_vllm.py:38]: add comment on why disable frontend multiprocessing?

File: presets/inference/text-generation/inference_api_vllm.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        "port": port
+    }
+    parser.set_defaults(**server_default_args)
+
+    # See https://docs.vllm.ai/en/latest/models/engine_args.html for more args
+    engine_default_args = {
+        "model": "/workspace/tfs/weights",
+        "cpu-offload-gb": 0,
+        "gpu-memory-utilization": 0.9,
+        "swap-space": 4,
```
Comment (ishaansehgal99) [inline presets/inference/text-generation/inference_api_vllm.py:48]: I noticed that `cpu-offload-gb`, `gpu-memory-utilization`, and `swap-space` are currently set to their default values based on the link.

open to making params like this configurable via environment variables in this/future PR? Something we'll need to think about how to do...

File: presets/inference/text-generation/inference_api_vllm.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        "disable_log_stats": False,
```
Comment (ishaansehgal99) [inline presets/inference/text-generation/inference_api_vllm.py:49]:
```suggestion
        "disable-log-stats": False,
```

File: presets/inference/text-generation/inference_api_vllm.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import logging
+import os
+import asyncio
+
+import uvloop
+from pydantic import BaseModel, Field
+from fastapi import Request, HTTPException
```
Comment (ishaansehgal99) [inline presets/inference/text-generation/inference_api_vllm.py:9]: Can we remove fastapi? Doesn't appear code is using it at all

File: presets/inference/text-generation/requirements.txt [hunk 10]
```diff
@@ -1,15 +1,17 @@
...
 accelerate==0.30.1
 fastapi>=0.111.0,<0.112.0  # Allow patch updates
-pydantic==2.7.4
+pydantic>=2.9
 uvicorn[standard]>=0.29.0,<0.30.0  # Allow patch updates
 peft==0.11.1
-numpy==1.22.4
+numpy
 sentencepiece==0.2.0
+jinja2>=3.1.0
```
Comment (ishaansehgal99) [inline presets/inference/text-generation/requirements.txt:14]: make separate requirements.txt file for vllm, lets not mess with something thats working. these pip packages have been pinned to very specific versions that are all compatible with each other

File: presets/inference/text-generation/tests/test_inference_api.py [hunk 11]
```diff
@@ -286,3 +286,72 @@ def test_generation_with_min_length(configured_app):
 
     assert min_new_tokens <= new_tokens <= max_new_tokens, "Response should generate at least min_new_tokens and at most max_new_tokens new tokens"
     assert len(total_tokens) <= max_length, "Total # of tokens has to be less than or equal to max_length"
+
+@pytest.fixture(params=[
+    {"model_path": "stanford-crfm/alias-gpt2-small-x21", "device": "cpu"},
+])
```
Comment (ishaansehgal99) [inline presets/inference/text-generation/tests/test_inference_api.py:292]: tests in seperate file for vllm?

File: presets/inference/text-generation/requirements.txt (hunk 10 above)
Comment (zhuangqh) [reply presets/inference/text-generation/requirements.txt:14]: Good point

Comment (zhuangqh) [reply presets/inference/text-generation/requirements.txt:14]: use different venv for each backend.

File: presets/inference/text-generation/inference_api_vllm.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (zhuangqh) [inline presets/inference/text-generation/inference_api_vllm.py:1]: add some test for it

File: Makefile [hunk 13]
```diff
@@ -101,9 +101,10 @@ unit-test: ## Run unit tests.
 ## E2E tests
 ## --------------------------------------
 
-inference-api-e2e: 
-	pip install -r presets/inference/text-generation/requirements.txt
-	pytest -o log_cli=true -o log_cli_level=INFO presets/inference/text-generation/tests
+inference-api-e2e:
+	pip install virtualenv
+	DEVICE=cpu ./hack/run-pytest-in-venv.sh presets/inference/vllm presets/inference/vllm/requirements.txt
```
Comment (zhuangqh) [inline Makefile:106]: skip the test on gh runner now. run this test in a gpu node offline

**Review by Fei-Guo — APPROVED** [review]

File: presets/inference/vllm/tests/test_inference_api.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
+import sys
```
Comment (ishaansehgal99) [inline presets/inference/vllm/tests/test_inference_api.py:1]: this is awesome that you got this working

File: presets/inference/vllm/tests/test_inference_api.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+    args = [
+        "python3",
+        os.path.join(parent_dir, "inference_api.py"),
+        "--model", TEST_MODEL,
+        "--chat-template", CHAT_TEMPLATE,
+        "--port", str(TEST_PORT)
+    ]
+    print(f">>> Starting server on port {TEST_PORT}")
+    process = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
Comment (ishaansehgal99) [inline presets/inference/vllm/tests/test_inference_api.py:38]: awesome you were able to run server in a subprocess sweet

File: presets/inference/vllm/tests/test_inference_api.py [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+# Get the parent directory of the current file
+parent_dir = str(Path(__file__).resolve().parent.parent)
+# Add the parent directory to sys.path
+sys.path.append(parent_dir)
+
+TEST_MODEL = "facebook/opt-125m"
+CHAT_TEMPLATE = ("{{ bos_token }}{% for message in messages %}{% if (message['role'] == 'user') %}"
+    "{{'<|user|>' + '\n' + message['content'] + '<|end|>' + '\n' + '<|assistant|>' + '\n'}}"
+    "{% elif (message['role'] == 'assistant') %}{{message['content'] + '<|end|>' + '\n'}}{% endif %}{% endfor %}")
```
Comment (ishaansehgal99) [inline presets/inference/vllm/tests/test_inference_api.py:19]: so in future pr - controller side - I believe we will have to have a default chat template for each model

**Review by ishaansehgal99 — APPROVED** [review]
sweeet

File: hack/run-pytest-in-venv.sh [hunk 17]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+
```
Comment (ishaansehgal99) [inline hack/run-pytest-in-venv.sh:2]: awesome

File: presets/inference/vllm/tests/test_inference_api.py (hunk 16 above)
Comment (zhuangqh) [reply presets/inference/vllm/tests/test_inference_api.py:19]: for those model which doesn't have a default chat template. we should provide one.

**Review by bangqipropel — APPROVED** [review]

**Review by ishaansehgal99 — APPROVED** [review]

**Pull request merged**